File: progect2/progect2/aether_cpp/src/evidence/smart_anti_boost_smoother.cpp
// SPDX-License-Identifier: LicenseRef-Aether3D-Proprietary
// Copyright (c) 2024-2026 Aether3D. All rights reserved.

#include "aether/evidence/smart_anti_boost_smoother.h"

#include <algorithm>
#include <cmath>
#include <limits>

namespace aether {
namespace evidence {

SmartAntiBoostSmoother::SmartAntiBoostSmoother(const SmartSmootherConfig& config)
    : config_(config) {
    config_.window_size = std::max<std::size_t>(1u, config_.window_size);
    config_.max_consecutive_invalid = std::max(1, config_.max_consecutive_invalid);
    config_.jitter_band = std::max(0.0, config_.jitter_band);
    history_.reserve(config_.window_size);
}

double SmartAntiBoostSmoother::handle_invalid_input() {
    consecutive_invalid_count_ += 1;
    if (consecutive_invalid_count_ >= config_.max_consecutive_invalid) {
        has_previous_smoothed_ = true;
        previous_smoothed_ = config_.worst_case_fallback;
        return config_.worst_case_fallback;
    }
    if (has_previous_smoothed_) {
        return previous_smoothed_;
    }
    return config_.worst_case_fallback;
}

double SmartAntiBoostSmoother::compute_median() const {
    if (history_.empty()) {
        return 0.0;
    }
    std::vector<double> sorted = history_;
    std::sort(sorted.begin(), sorted.end());
    const std::size_t n = sorted.size();
    if ((n % 2u) == 0u) {
        return 0.5 * (sorted[n / 2u - 1u] + sorted[n / 2u]);
    }
    return sorted[n / 2u];
}

double SmartAntiBoostSmoother::compute_smoothed(double new_value) {
    const double median = compute_median();
    const double previous = has_previous_smoothed_ ? previous_smoothed_ : median;
    const double change = new_value - previous;

    if (std::fabs(change) < config_.jitter_band) {
        return median;
    }

    if (change > 0.0) {
        if (change > config_.jitter_band * 3.0) {
            return previous + change * config_.anti_boost_factor;
        }
        return previous + change * config_.normal_improve_factor;
    }

    // In capture mode, degrade_factor is forced to 0: no visual regression.
    const double effective_degrade = config_.capture_mode ? 0.0 : config_.degrade_factor;
    return previous + change * effective_degrade;
}

double SmartAntiBoostSmoother::add(double value) {
    if (!std::isfinite(value)) {
        return handle_invalid_input();
    }

    consecutive_invalid_count_ = 0;

    if (history_.size() >= config_.window_size) {
        history_.erase(history_.begin());
    }
    history_.push_back(value);

    double smoothed = compute_smoothed(value);

    // In capture mode, enforce high-water mark: output never falls below peak.
    if (config_.capture_mode) {
        smoothed = std::max(smoothed, high_water_mark_);
        high_water_mark_ = smoothed;
    }

    has_previous_smoothed_ = true;
    previous_smoothed_ = smoothed;
    return smoothed;
}

void SmartAntiBoostSmoother::reset() {
    history_.clear();
    has_previous_smoothed_ = false;
    previous_smoothed_ = 0.0;
    consecutive_invalid_count_ = 0;
    high_water_mark_ = 0.0;
}

}  // namespace evidence
}  // namespace aether

File: progect2/progect2/Core/Evidence/Smoothing/SmartAntiBoostSmoother.swift
// SPDX-License-Identifier: LicenseRef-Aether3D-Proprietary
// Copyright (c) 2024-2026 Aether3D. All rights reserved.

//
// SmartAntiBoostSmoother.swift
// Aether3D
//
// PR3 - Smart Anti-Boost Dual Channel Smoother
// Conditional anti-boost, not punishment
//

import Foundation
#if canImport(CAetherNativeBridge)
import CAetherNativeBridge
#endif

/// Smart Anti-Boost Dual Channel Smoother
///
/// DESIGN:
/// - Anti-boost ONLY when jump exceeds jitterBand (suspicious)
/// - Normal improvement: faster recovery (configurable)
/// - K consecutive invalid: force worst-case fallback
/// - Hysteresis: prevent oscillation at boundaries
public final class SmartAntiBoostSmoother: @unchecked Sendable {

    // ═══════════════════════════════════════════════════════════════════════
    // MARK: - State
    // ═══════════════════════════════════════════════════════════════════════

    private let config: SmootherConfig
    private let nativeHandle: OpaquePointer?

    /// Previous smoothed value (for change detection)
    private var previousSmoothed: Double?

    /// Consecutive invalid frame counter
    private var consecutiveInvalidCount: Int = 0

    // ═══════════════════════════════════════════════════════════════════════
    // MARK: - Initialization
    // ═══════════════════════════════════════════════════════════════════════

    /// Initialize smoother
    ///
    /// - Parameters:
    ///   - windowSize: Window size for median computation
    ///   - config: Configuration (default if not specified)
    public init(windowSize: Int = 5, config: SmootherConfig = .default) {
        let resolvedWindowSize = max(1, windowSize)
        self.config = config
#if canImport(CAetherNativeBridge)
        var nativeConfig = aether_smart_smoother_config_t(
            window_size: Int32(resolvedWindowSize),
            jitter_band: config.jitterBand,
            anti_boost_factor: config.antiBoostFactor,
            normal_improve_factor: config.normalImproveFactor,
            degrade_factor: config.degradeFactor,
            max_consecutive_invalid: Int32(max(1, config.maxConsecutiveInvalid)),
            worst_case_fallback: config.worstCaseFallback,
            capture_mode: 0
        )
        var handle: OpaquePointer?
        let rc = aether_smart_smoother_create(&nativeConfig, &handle)
        self.nativeHandle = (rc == 0) ? handle : nil
#else
        self.nativeHandle = nil
#endif
    }

    deinit {
#if canImport(CAetherNativeBridge)
        if let nativeHandle {
            _ = aether_smart_smoother_destroy(nativeHandle)
        }
#endif
    }

    // ═══════════════════════════════════════════════════════════════════════
    // MARK: - Core API
    // ═══════════════════════════════════════════════════════════════════════

    /// Add value and return smoothed result
    ///
    /// BEHAVIOR:
    /// 1. If invalid (NaN/Inf): increment invalidCount, return previous or fallback
    /// 2. If valid: reset invalidCount, add to history, compute smoothed
    /// 3. If consecutiveInvalidCount >= maxConsecutiveInvalid: return worst-case
    ///
    /// - Parameter value: Input value
    /// - Returns: Smoothed value
    public func addAndSmooth(_ value: Double) -> Double {
#if canImport(CAetherNativeBridge)
        if let nativeHandle {
            var smoothed = 0.0
            if aether_smart_smoother_add(nativeHandle, value, &smoothed) == 0 {
                return smoothed
            }
        }
#endif
        // Fail-closed bridge behavior when native kernel is unavailable.
        guard value.isFinite else {
            return handleInvalidInput()
        }
        consecutiveInvalidCount = 0
        previousSmoothed = value
        return value
    }

    // ═══════════════════════════════════════════════════════════════════════
    // MARK: - Internal Logic
    // ═══════════════════════════════════════════════════════════════════════

    private func handleInvalidInput() -> Double {
        consecutiveInvalidCount += 1

        // If too many consecutive invalid, force worst-case
        if consecutiveInvalidCount >= config.maxConsecutiveInvalid {
            previousSmoothed = config.worstCaseFallback
            return config.worstCaseFallback
        }

        // Otherwise return previous smoothed (or worst-case if none)
        return previousSmoothed ?? config.worstCaseFallback
    }

    /// Reset all state
    public func reset() {
#if canImport(CAetherNativeBridge)
        if let nativeHandle {
            _ = aether_smart_smoother_reset(nativeHandle)
        }
#endif
        previousSmoothed = nil
        consecutiveInvalidCount = 0
    }
}
